source/gui/accPropServer.py

```diff
@@ -9,29 +9,53 @@
 from logHandler import log
 from  comtypes.automation import VT_EMPTY
 from  comtypes import COMObject
-from comInterfaces.Accessibility import IAccPropServer
-from abc import ABCMeta, abstractmethod
+from comInterfaces.Accessibility import IAccPropServer, ANNO_CONTAINER, ANNO_THIS
+from abc import ABCMeta, abstractmethod, abstractproperty
 from six import with_metaclass
 import weakref
+import winUser
+import wx
 
 class IAccPropServer_Impl(with_metaclass(ABCMeta, COMObject)):
-	"""Base class for implementing a COM interface for AccPropServer.
-	Please override the _GetPropValue method, not GetPropValue.
-	GetPropValue wraps _getPropValue to catch and log exceptions (Which for some reason NVDA's logger misses when they occur in GetPropValue).
+	"""Base class for implementing a COM interface for a hwnd based AccPropServer\
+	to annotate a WX control.
+	The AccPropServer registers itself using the window handle of the WX control.
+	When the WX control is destroyed, the instance is automatically unregistered.
+	This should eventually be dropped in favor of WX' own annotation support,
+	blocked by wxWidgets/Phoenix#1129.
+	Please override the L{_GetPropValue} method, not L{GetPropValue}.
+	L{GetPropValue} wraps L{_getPropValue} to catch and log exceptions (Which for some reason NVDA's logger misses when they occur in GetPropValue).
+	You must also provide the L{properties} property.
 	"""
 
 	_com_interfaces_ = [
 		IAccPropServer
 	]
 
-	def __init__(self, control, *args, **kwargs):
+	def __init__(self, control, annotateChildren=True):
 		"""Initialize the instance of AccPropServer. 
 		@param control: the WX control instance, so you can look up things in the _getPropValue method.
 			It's available on self.control.
 		@Type control: Subclass of wx.Window
+		@param annotateChildren: whether the WX control is a container which children should be annotated.
+		@type annotateChildren: bool
 		"""
 		self.control = weakref.ref(control)
-		super(IAccPropServer_Impl, self).__init__(*args, **kwargs)
+		self.hwnd = control.GetHandle()
+		super(IAccPropServer_Impl, self).__init__()
+		# Import late to avoid circular import
+		from IAccessibleHandler import accPropServices
+		accPropServices.SetHwndPropServer(
+			hwnd=self.hwnd,
+			idObject=winUser.OBJID_CLIENT,
+			idChild=0,
+			paProps=self.properties,
+			cProps=len(self.properties),
+			pServer=self,
+			AnnoScope=ANNO_CONTAINER if annotateChildren else ANNO_THIS
+		)
+		# clean up of the accPropServices needs to happen when the control is destroyed.
```

Comment:
I think it's helpful for a reader to understand why we must use bind rather than alternatives.

```suggestion
		# clean up of accPropServices needs to happen when the control is destroyed. We can't rely on 
		# pythons `__del__` method to be called, and the wx framework does not call Destroy on child controls,
		# automatically. Instead we can bind to the "window destroy" event for the control to do necessary 
		# cleanup (wxWidgets/Phoenix/#630). Not performing this cleanup results in a reference to the parent
		# window, keeping it from being deleted correctly. This can cause a freeze on exit of NVDA.
```